Cython/Compiler/ModuleNode.py

```diff
@@ -1764,10 +1764,9 @@ def generate_dealloc_function(self, scope, code):
             if needs_gc:
                 # The base class deallocator probably expects this to be tracked,
                 # so undo the untracking above.
-                if base_type.scope:
-                    # Assume that we know whether the base class uses GC or not.
-                    if base_type.scope.needs_gc():
-                        code.putln("PyObject_GC_Track(o);")
+                # Assume that we do not know whether the base class uses GC or not.
+                if base_type.scope and base_type.scope.needs_gc():
+                    code.putln("PyObject_GC_Track(o);")
                 else:
                     code.putln("if (PyType_IS_GC(%s)) PyObject_GC_Track(o);" % base_cname)
```

Comment:
```suggestion
                # extern types are unknown so we always have to check for gc
                if not base_type.scope or (base.type.is_external and not base_type.is_builtin_type):
                    code.putln("if (PyType_IS_GC(%s)) PyObject_GC_Track(o);" % base_cname)
                else:
                    # Assume that we know whether the base class uses GC or not.
                    if base_type.scope.needs_gc():
                        code.putln("PyObject_GC_Track(o);")
```

I think @scoder's suggestion was something like this, to special case extern types and treat everything else as not needing GC.

I've reordered the if/else because the logic was clearer to me that way, but you may not want to do this.

Reply:
Let me import the change and see if it works in our codebase